### PR TITLE
[SPARK-26798][SQL] HandleNullInputsForUDF should trust nullability

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

There is a very old TODO in `HandleNullInputsForUDF`, saying that we can skip the null check if input is not nullable. We leverage the nullability info at many places, we can trust it here too.

## How was this patch tested?

re-enable an ignored test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @gatorsmile @maryannxue

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -2147,27 +2147,34 @@ class Analyzer(
...
-          // prevent this rule from being applied repeatedly.
-          val newInputsNullSafe = inputsNullSafe.map(_ => true)
-          inputsNullCheck
-            .map(If(_, Literal.create(null, udf.dataType),
-              udf.copy(inputsNullSafe = newInputsNullSafe)))
-            .getOrElse(udf)
+          assert(primitiveInputs.length == inputs.length)
+
+          val inputNullCheck = primitiveInputs.zip(inputs).collect {
+            case (isPrimitive, input) if isPrimitive && input.nullable => IsNull(input)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: This is the major change made in this PR.

**Comment by maropu** [comment]
We need to reverse the `ScalaUDF` flag in `FileFormatDataWriter` and `UDFRegistration`?

https://github.com/apache/spark/blob/196ca0c8f5f5caa2ae588ff4a63ce314aa42aecc/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatDataWriter.scala#L184

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala#L722

**Comment by gatorsmile** [comment]
It sounds like we do not have the test cases to cover it?

**Comment by maropu** [comment]
Yea, I think so. If possible, better to add them.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -2147,27 +2147,36 @@ class Analyzer(
...
+          }.reduceLeftOption[Expression](Or)
+
+          if (inputNullCheck.isDefined) {
+            // Once we add an `If` check above the udf, it is safe to mark those checked inputs
+            // as null-safe (i.e., wrap with `KnownNotNull`), because the null-returning
+            // branch of `If` will be called if any of these checked inputs is null. Thus we can
+            // prevent this rule from being applied repeatedly.
+            val newInputs = inputPrimitivesPair.map {
+              case (isPrimitive, input) =>
+                if (isPrimitive && input.nullable) {
```
Comment (maryannxue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Does it mean we trust the nullability from the source here?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: yes, and we leverage nullability at many places already. Like the `NullPropagation` rule.

Comment (maryannxue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: This is just to confirm. And I think this is the write way to go.

**Comment by maropu** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 3]
```diff
@@ -31,9 +31,9 @@ import org.apache.spark.sql.types.{AbstractDataType, DataType}
  *                  null. Use boxed type or [[Option]] if you wanna do the null-handling yourself.
  * @param dataType  Return type of function.
  * @param children  The input expressions of this UDF.
- * @param inputsNullSafe Whether the inputs are of non-primitive types or not nullable. Null values
- *                       of Scala primitive types will be converted to the type's default value and
- *                       lead to wrong results, thus need special handling before calling the UDF.
+ * @param inputPrimitives Whether the inputs are of primitive types. Null values of Java primitive
```
Comment (maryannxue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: => The analyzer should be aware of Scala primitive types so as to make the UDF return null if there is any null input value of these types. On the other hand, Java UDFs can only have boxed types, thus this parameter will always be all false.

**Comment by maryannxue** [comment]
LGTM except one minor suggestion.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala (hunk 3 above)
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: +1

**Comment by maropu** [comment]
retest this please

**Review by maropu — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
@HyukjinKwon Sorry to bother you, but I probably failed to merge this pr into master.
It seems github and gitbox have different commit logs? Could you check?
- https://gitbox.apache.org/repos/asf?p=spark.git;a=shortlog;h=refs/heads/master
- https://github.com/apache/spark/commits/master

**Comment by gatorsmile** [comment]
Done. We need to merge it again. Please use gitbox @maropu Thanks!

**Comment by cloud-fan** [comment]
according to https://github.com/apache/spark-website/commit/92606b2e7849b9d743ef2a8176438142420a83e5#diff-17faa4bab13b7530a3e1b627bb798ad0, we should use github? cc @HyukjinKwon @srowen

**Comment by maropu** [comment]
@gatorsmile Thanks! That's a relief! Either way, I think we need to explicitly write "Don't push into `[github|gitbox]` in spark-website.

**Comment by srowen** [comment]
Oh I see this was the commit that got lost between gitbox/github. Again I apologize that we didn't make extra clear that we should use one repo, github, even though we can push to both. It is on the website, FWIW. @maropu you have updated your remotes locally, and can reopen a PR with this change?

**Comment by maropu** [comment]
It seems most committers use `apache` based on spark-website? I'll update my remote, and then try to merge this pr into master. Thanks!

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by HyukjinKwon** [comment]
Oops, I completely missed this thread. Sorry I noticed this one late :(.

**Comment by maropu** [comment]
Never mind, @HyukjinKwon! Thanks!

**Pull request closed without merging**